src/main/java/org/junitpioneer/jupiter/RangeSourceProvider.java

```diff
@@ -0,0 +1,97 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.ArgumentsProvider;
+import org.junit.jupiter.params.provider.ArgumentsSource;
+import org.junit.platform.commons.util.Preconditions;
+
+/**
+ * Provides a range of {@link Number}s, as defined by an annotation which is its {@link ArgumentsSource}.
+ * Such an annotation should have the following properties:
+ * <ul>
+ *     <li>{@code from} a primitive value for the "start" of the range.</li>
+ *     <li>{@code to} a primitive value for the "end" of the range. {@code to} must have the same type as {@code from}.</li>
+ *     <li>{@code step} a primitive value for the difference between each two values of the range.</li>
+ *     <li>{@code closed} a {@code boolean} value describing if the range includes the last value (cloded), or not (open).</li>
+ * </ul>
+ *
+ * @see IntRangeSource
+ * @see ShortRangeSource
+ * @see LongRangeSource
+ * @see ByteRangeSource
+ * @see DoubleRangeSource
+ * @see FloatRangeSource
+ */
+public class RangeSourceProvider implements ArgumentsProvider {
+
+	@Override
+	public Stream<? extends Arguments> provideArguments(ExtensionContext context) throws Exception {
+		// @formatter:off
+		// since it's a method annotation, the element will always be present
+		List<Annotation> argumentsSources = context
+				.getElement()
+				.map(method -> Arrays
+						.stream(method.getAnnotations())
+						.filter(annotations -> Arrays
+								.stream(annotations.annotationType().getAnnotationsByType(ArgumentsSource.class))
+								.anyMatch(annotation -> getClass().equals(annotation.value())))
+						.collect(Collectors.toList()))
+				.get();
+
+		Preconditions.condition(
+				argumentsSources.size() == 1,
+				() -> String.format(
+						"Expected exactly one annotation to provide an ArgumentSource, found %d.",
+						argumentsSources.size()));
+		// @formatter:on
+
+		Annotation argumentsSource = argumentsSources.get(0);
+		Class<? extends Annotation> argumentsSourceClass = argumentsSource.annotationType();
+
+		Method stepMethod = argumentsSourceClass.getMethod("step");
+		Number stepValue = (Number) stepMethod.invoke(argumentsSource);
+
+		Method fromMethod = argumentsSourceClass.getMethod("from");
+		Number fromValue = (Number) fromMethod.invoke(argumentsSource);
+
+		Method toMethod = argumentsSourceClass.getMethod("to");
+		Number toValue = (Number) toMethod.invoke(argumentsSource);
+
+		Method closeMethod = argumentsSourceClass.getMethod("closed");
+		Boolean closedValue = (Boolean) closeMethod.invoke(argumentsSource);
+
+		// @formatter:off
+		Preconditions.condition(
+				fromMethod.getReturnType() == toMethod.getReturnType(),
+				() -> String.format(
+						"Illegal range. Different classes of from (%s) and to (%s).",
+						fromMethod.getReturnType(), toMethod.getReturnType()));
```

Comment:
It's a remnant of one of the previous iterations. The idea wat to prevent any future contributor implementing a range with conflicting `from` and `to`, but that's probably a mute point now. I'll remove it.